tensorflow/core/kernels/resource_variable_ops.cc

```diff
@@ -214,17 +219,61 @@ void ReadVariablesOp::Compute(OpKernelContext* ctx) {
   }
 }
 
+void ReadVariableWithoutCopyOp::Compute(OpKernelContext* ctx) {
+  core::RefCountPtr<Var> variable;
+  const ResourceHandle& handle = HandleFromInput(ctx, 0);
+  const auto status = LookupResource(ctx, handle, &variable);
+  bool mode_conversion = false;
+  OP_REQUIRES(ctx, status.ok(),
+              errors::FailedPrecondition(
+                  "Could not find variable ", handle.name(), ". ",
+                  "This could mean that the variable has been deleted. ",
+                  "In TF1, it can also mean the variable is uninitialized. ",
+                  "Debug info: container=", handle.container(),
+                  ", status error message=", status.error_message()));
+
+  if (variable->copy_on_read_mode.load()) {
+    // If the variable is in copy-on-read mode, obtain an exclusive
+    // lock on the variable before reading it without making a copy
+    mutex_lock ml(*variable->mu());
+    const Tensor* t = variable->tensor();
+    OP_REQUIRES(
+        ctx, dtype_ == t->dtype(),
+        errors::InvalidArgument(
+            "Trying to read variable with wrong dtype. Expected ",
+            DataTypeString(dtype_), " got ", DataTypeString(t->dtype())));
+    ctx->set_output(0, *t);
```

Comment:
The output tensor will be unstable. When the COR variable's buffer got modified, the output tensor's value will silently change, which breaks TF tensors' immutability guarantee.

My preferred solution is to add a `TurnOffCopyOnRead` (or `DisableCopyOnRead`) op to turn off a variable's copy_on_read_mode.